### ISSUE TITLE
Race conditions running Orchard migrations on multiple instances

### ISSUE THREAD

**Issue by jersiovic** [OrchardCMS/OrchardCore#7941]
Hello

We are facing race conditions on migration when we use two instances in an Azure App Service with Container.

We receive exceptions cause another instance executed a migration and the changes on migrations are attempeted to be executed by both instances. This causes problems cause one of the two instances didn't complete its initialization.

Taking a look at the code we saw Orchard uses SemaphoreSlim to avoid two threads excuting at the same time migrations for a tenant. Is there any artifact like TaskLease at Orchard 1? Should it be used to avoid problems on the scenario I mention?

**Comment by jtkech** [comment]
We have an `IDistributedLock` that we don't use yet, the default implementation is a `LocalLock` based on multiple `SemaphoreSlim` per tenant, and we have a concrete implementation based on Redis.

The interface allows to `TryAcquireLockAsync()` a named lock (for a given tenant) in a given timeout and that will be valid for a given expiration, allowing to do e.g. atomic operations in a distributed env. There is also an `IsLockAcquiredAsync()` allowing to do exclusive tasks by just doing nothing if a named lock is already acquired.

Maybe we will use it e.g. for Workflow Singleton in a distributed env, Exclusive BackgroundTask ...

About migrations, when they are executed on a fresh setup, normally it is on only one instance. Through the admin UI, by enabling features, executing recipes, yes it may happen concurrently. When the code is updated with some migration steps, yes it may happen concurrently if multiples instances are starting, is it your scenario?

So, maybe we will need to also use an `IDistributedLock` for migrations, but maybe only if it is not during a fresh setup, hmm but in this case the lock resolved would be a simple `LocalLock`, hmm maybe not at some point if you enable `RedisLock` in the setup recipe. Anyway, i think it is better to not use a lock on setup (`TenantState.Initializing`), but yes maybe we have to use it on migrations if the tenant is already `TenantState.Running`.

> We are facing race conditions on migration when we use two instances

Does it throws exceptions? Or just let the 2 instances not synchronized.

**Comment by jersiovic** [comment]
> We have an `IDistributedLock` that we don't use yet, the default implementation is a `LocalLock` based on multiple `SemaphoreSlim` per tenant, and we have a concrete implementation based on Redis.
> 
> The interface allows to `TryAcquireLockAsync()` a named lock (for a given tenant) in a given timeout and that will be valid for a given expiration, allowing to do e.g. atomic operations in a distributed env. There is also an `IsLockAcquiredAsync()` allowing to do exclusive tasks by just doing nothing if a named lock is already acquired.
> 
> Maybe we will use it e.g. for Workflow Singleton in a distributed env, Exclusive BackgroundTask ...
> 
> About migrations, when they are executed on a fresh setup, normally it is on only one instance. Through the admin UI, by enabling features, executing recipes, yes it may happen concurrently. When the code is updated with some migration steps, yes it may happen concurrently if multiples instances are starting, is this last case your scenario?
> 
Yes, that is our scenario. We update the container of the webapp with a new docker imame that contain new version of our custom modules. That makes both instances to execute pending migrations of updated modules

> So, maybe we will need to also use an `IDistributedLock` for migrations, but maybe only if it is not during a fresh setup, hmm but in this case the lock resolved would be a simple `LocalLock`, hmm maybe not at some point if you enable `RedisLock` in the setup recipe. Anyway, i think it is better to not use a lock on setup (`TenantState.Initializing`), but yes maybe we have to use it on migrations (when activating the tenant) if the tenant is already `TenantState.Running`.
> 
> > We are facing race conditions on migration when we use two instances
> 
> Does it throws exceptions? Or just let the 2 instances not synchronized.
It throws the exception on migrations cause an alter table removing a column cannot be executed because a previous instance did the job. This exception is logged. The instance which throws the exception doesn't work properly. It makes in one request with the browser everything works properly if it is answered by the instance without problems but on  next request if the instance which answer the request is the one which throw the exception things doesn't work properly.

**Comment by jtkech** [comment]
@jersiovic 

Okay, so i think for now there is nothing to do in the source code, just if we provide a new migration step in the future, and people will have to do the same for their own modules.

I will show you asap how to do it with the existing `IDistributedLock` based on redis, so that only one instance will do a migration step, and other instances will have to wait before continuing their own execution flow.

But you will need a Redis instance in your container, and enable the `RedisLock` feature.

**Comment by jersiovic** [comment]
Cool, thank you

**Comment by jtkech** [comment]
@jersiovic 

Look at #8161 where we use a distributed lock while activating a given shell (where migration steps are executed)

Can you try it when you will have time? And see if it fixes your issue

You will need a Redis instance, an `OrchardCore_Redis:Configuration` string as commented in our appsettings.json

And then enable the `RedisLock` feature

**Comment by jersiovic** [comment]
Cool thank you @jtkech
